### PR TITLE
Support App Signing by Google Play 

### PR DESCRIPTION
**Pull request by wkozyra95**
Added two commands
- `opt-in-google-play-signing`  simplifies the process of migrating into google play app signing(series of instruction user needs to follow)
- `fetch:android:upload-cert` - extracting certificate from currently used keystore

### CODE REVIEW

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+  async exportPrivateKey() {
+    log(
+      `Go to "App signing" tab in Google Play console, select "${chalk.bold(
+        'Export and upload a key (not using a Java keystore)'
+      )}" and copy encryption key that is listed in step 1.`
+    );
+    const encryptKeyQuestion = [
+      { type: 'input', name: 'encryptionKey', message: 'Google Play encryption key' },
+    ];
+    const { encryptionKey } = await prompt(encryptKeyQuestion);
```
Comment (dsokal) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]: `prompt` is just a wrapper for https://github.com/enquirer/enquirer You can avoid the following `if` by passing `validate` function in the question object.

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+      log
+    );
+    log('Upload certificate');
+    await Credentials.Android.logKeystoreHashes(
+      { keystorePath: this.uploadKeystore, ...this.uploadKeystoreCredentials },
+      log
+    );
+
+    log.warn(
+      `On previously opened Google Play console page, upload ${chalk.underline(
```
Comment (dsokal) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      `On the previously opened Google Play console page, upload ${chalk.underline(
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        this.privateSigningKey
+      )} as "${chalk.bold('APP SIGNING PRIVATE KEY')}" and ${chalk.underline(
+        this.publicUploadCert
+      )} as "${chalk.bold('UPLOAD KEY PUBLIC CERTIFICATE')}"`
+    );
+
+    log.warn(
+      `Next step will ${chalk.red(
+        'remove current keystore from Expo servers'
```
Comment (dsokal) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
        'remove your current keystore from Expo servers'
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+      )}, make sure that private key is uploaded successfully and compare hashes displayed above with ones in console.`
```
Comment (dsokal) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      )}, make sure that private key is uploaded successfully and compare hashes displayed above with the ones printed in console.`
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        platform: 'android',
+        username,
+        experienceName: `@${username}/${exp.slug}`,
+      }
+    );
+
+    log(
+      `Original keystore is stored in ${
+        this.signKeystore
+      }, remove it only if you are sure that Google Play signing is enabled on your app`
```
Comment (dsokal) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      }, remove it only if you are sure that Google Play App Signing is enabled for your app.`
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+  async runWithoutUploadCert() {
+    log('App signing certificate');
+    await Credentials.Android.logKeystoreHashes(
+      { keystorePath: this.signKeystore, ...this.signKeystoreCredentials },
+      log
+    );
+
+    log.warn(
+      `On previously opened Google Play console page, upload ${chalk.underline(
```
Comment (dsokal) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      `On the previously opened Google Play console page, upload ${chalk.underline(
```

File: packages/xdl/src/credentials/AndroidCredentials.js [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+  outputPath: string,
+  log: any = logger.info.bind(logger)
+) {
+  const encryptToolPath = path.join(UserSettings.dotExpoHomeDirectory(), 'android_tools_pepk.jar');
+  if (!fs.existsSync(encryptToolPath)) {
+    log(`Downloading PEPK tool from Google Play to ${encryptToolPath}`);
+    const downloadUrl =
+      'https://www.gstatic.com/play-apps-publisher-rapid/signing-tool/prod/pepk.jar';
+    const file = fs.createWriteStream(encryptToolPath);
+    const response = await axios({ url: downloadUrl, method: 'GET', responseType: 'stream' });
```
Comment (dsokal) [inline packages/xdl/src/credentials/AndroidCredentials.js]: Could you please add a progress bar here?

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+    const { confirm: confirmEnabled } = await prompt(confirmQuestion);
+    if (confirmEnabled) {
+      log('Google Play signing already enabled, there is nothing to do here.');
+      process.exit(0);
+    }
+
+    this.signKeystore = path.join(this.projectDir, `${slug}_sign.jks.bak`);
+    this.uploadKeystore = path.join(this.projectDir, `${slug}_upload.jks.tmp`);
+    this.privateSigningKey = path.join(this.projectDir, `${slug}_private_sign_key`);
+    this.publicUploadCert = path.join(this.projectDir, `${slug}_upload_cert.pem`);
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]: Those values are not passed anywhere as a group and usually are used separately, so I don't see we could gain anything by wrapping them into an object.

File: packages/expo-cli/src/commands/fetch/index.js [hunk 9]
```diff
@@ -19,7 +23,14 @@ export default program => {
   program
     .command('fetch:android:hashes [project-dir]')
     .description(
-      "Fetch this project's Android key hashes needed to setup Google/Facebook authentication."
+      "Fetch this project's Android key hashes needed to setup Google/Facebook authentication. If you are using Google Play signing this app will be signed with difrent key after publishing to the store, in that case you need to use hashes displayed in Google Play console"
```
Comment (ide) [inline packages/expo-cli/src/commands/fetch/index.js]:
```suggestion
      "Fetch this project's Android key hashes needed to set up Google/Facebook authentication. Note: if you are using Google Play signing, this app will be signed with a different key after publishing to the store, and you'll need to use the hashes displayed in the Google Play console."
```

File: packages/expo-cli/src/commands/fetch/index.js [hunk 10]
```diff
@@ -19,7 +23,14 @@ export default program => {
...
     )
     .asyncActionProjectDir(fetchAndroidHashesAsync, true);
+
+  program
+    .command('fetch:android:upload-cert [project-dir]')
+    .description(
+      "Fetch this project's upload certificate needed after opting into App Signing by Google Play or after reseting previous upload certifacate."
```
Comment (ide) [inline packages/expo-cli/src/commands/fetch/index.js]:
```suggestion
      "Fetch this project's upload certificate needed after opting in to app signing by Google Play or after resetting a previous upload certificate."
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+    } catch (error) {
+      log.error(error);
+      await this.cleanup(true);
+    }
+    await this.afterStoreSubmit(username, exp);
+  }
+
+  async init(slug: string) {
+    log.warn(
+      'Make sure you are not using Google Play App Signing already, this process will remove your current keystore from Expo servers.'
```
Comment (ide) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      'Make sure you are not using Google Play App Signing already as this process will remove your current keystore from Expo servers.'
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+    );
+    log(
+      `You can check it here ${chalk.underline(
```
Comment (ide) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      `You can check whether you are using Google Play App Signing here: ${chalk.underline(
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+        'https://play.google.com/apps/publish'
+      )}, select app and go to \"Release managment\" -> \"App signing\" tab, if you are already using Google Play App Signing, there will be a message \"App Signing by Google Play is enabled for this app\" at the top of the page.`
```
Comment (ide) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      )}. Select your app and go to "Release management" → "App signing" tab. If you are already using Google Play App Signing, there will be a message that says, "App Signing by Google Play is enabled for this app," at the top of the page.`
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+    const confirmQuestion = [
+      {
+        type: 'confirm',
+        name: 'confirm',
+        message: 'Is Google Play App Signing enabled for this app?',
+      },
+    ];
+    const { confirm: confirmEnabled } = await prompt(confirmQuestion);
+    if (confirmEnabled) {
+      log('Google Play App Signing is already enabled, there is nothing to do here.');
```
Comment (ide) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      log('Google Play App Signing is already enabled; there is nothing to do here.');
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+    this.signKeystore = path.join(this.projectDir, `${slug}_sign.jks.bak`);
+    this.uploadKeystore = path.join(this.projectDir, `${slug}_upload.jks.tmp`);
+    this.privateSigningKey = path.join(this.projectDir, `${slug}_private_sign_key`);
+    this.publicUploadCert = path.join(this.projectDir, `${slug}_upload_cert.pem`);
+    await this.cleanup(true);
+  }
+
+  async exportPrivateKey() {
+    log(
+      `Go to "App signing" tab in Google Play console, select "${chalk.bold(
```
Comment (ide) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      `Go to the "App signing" tab in the Google Play console, select "${chalk.bold(
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+        'Export and upload a key (not using a Java keystore)'
+      )}" and copy encryption key that is listed in step 1.`
```
Comment (ide) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      )}" and copy the encryption key that is listed in step 1.`
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+    );
+    const encryptKeyQuestion = [
+      {
+        type: 'input',
+        name: 'encryptionKey',
+        message: 'Google Play encryption key',
+        validate: value =>
+          (value.length === 136 && /^[A-Fa-f0-9]+$/.test(value)) ||
+          'Encryption key needs to be a hex encoded 68 byte string (a 4-byte identity followed by a 64-byte P256 point).',
```
Comment (ide) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
          'Encryption key needs to be a hex-encoded 68-byte string (a 4-byte identity followed by a 64-byte P-256 point).',
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+        this.privateSigningKey
+      )} as "${chalk.bold('APP SIGNING PRIVATE KEY')}" and ${chalk.underline(
+        this.publicUploadCert
+      )} as "${chalk.bold('UPLOAD KEY PUBLIC CERTIFICATE')}"`
+    );
+
+    log.warn(
+      `Next step will ${chalk.red(
+        'remove your current keystore from Expo servers'
+      )}, make sure that private key is uploaded successfully and compare hashes displayed above with the ones printed in console.`
```
Comment (ide) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      )}. Make sure that private key is uploaded successfully and compare the hashes displayed above with the ones printed in the console.`
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+    log.warn(
+      `On the previously opened Google Play console page, upload ${chalk.underline(
+        this.privateSigningKey
+      )} as "${chalk.bold('APP SIGNING PRIVATE KEY')}" and ${chalk.underline(
+        this.publicUploadCert
+      )} as "${chalk.bold('UPLOAD KEY PUBLIC CERTIFICATE')}"`
+    );
+
+    log.warn(
+      `Next step will ${chalk.red(
```
Comment (ide) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      `The next step will ${chalk.red(
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+      },
+      {
+        platform: 'android',
+        username,
+        experienceName: `@${username}/${exp.slug}`,
+      }
+    );
+
+    log(
+      `Original keystore is stored in ${
```
Comment (ide) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      `The original keystore is stored in ${
```

File: packages/expo-cli/src/commands/google-play/AppSigningOptIn.js [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+        this.signKeystore
+      }, remove it only if you are sure that Google Play App Signing is enabled for your app.`
```
Comment (ide) [inline packages/expo-cli/src/commands/google-play/AppSigningOptIn.js]:
```suggestion
      }; remove it only if you are sure that Google Play App Signing is enabled for your app.`
```

File: packages/expo-cli/src/commands/opt-into-google-play-signing.js [hunk 22]
```diff
@@ -0,0 +1,15 @@
+/* @flow */
+
+import AppSignigOptInProcess from './google-play/AppSigningOptIn';
```
Comment (ide) [inline packages/expo-cli/src/commands/opt-into-google-play-signing.js]:
```suggestion
import AppSigningOptInProcess from './google-play/AppSigningOptIn';
```

File: packages/expo-cli/src/commands/opt-into-google-play-signing.js [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+
+export default (program: any) => {
+  program
+    .command('opt-in-google-play-signing [project-dir]')
+    .description(
+      'Switch from old method of signing APKs to App Signing by Google Play. APK will be signed with upload key and after uploading to store signature will be stripped and app will be signed with key from original keystore.'
```
Comment (ide) [inline packages/expo-cli/src/commands/opt-into-google-play-signing.js]:
```suggestion
      'Switch from the old method of signing APKs to the new App Signing by Google Play. The APK will be signed with an upload key and after uploading to it to the store, app will be re-signed with the key from the original keystore.'
```

File: packages/xdl/src/credentials/AndroidCredentials.js [hunk 24]
```diff
@@ -0,0 +1,255 @@
...
+    throw new Error('Unable to fetch credentials for this project. Are you sure they exist?');
+  }
+  const { keystore, keystorePassword, keystoreAlias: keyAlias, keyPassword } = credentials;
+
+  const storeBuf = Buffer.from(keystore, 'base64');
+  log(`Writing keystore to ${outputPath}...`);
+  fs.writeFileSync(outputPath, storeBuf);
+  if (logSecrets) {
+    log('Done writing keystore to disk.');
+    log(`Save these important values as well:
```
Comment (ide) [inline packages/xdl/src/credentials/AndroidCredentials.js]: Highlight this section with chalk so the developer knows they need to pay attention to this message.

File: packages/xdl/src/credentials/AndroidCredentials.js [hunk 25]
```diff
@@ -0,0 +1,255 @@
...
+
+  Keystore password: ${chalk.bold(keystorePassword)}
+  Key alias:         ${chalk.bold(keyAlias)}
+  Key password:      ${chalk.bold(keyPassword)}
+  `);
+    log('Keystore saved!');
```
Comment (ide) [inline packages/xdl/src/credentials/AndroidCredentials.js]:
```suggestion
```
"Keystore saved" sounds like we might have saved the information above (ex: password, alias). Just remove this message so we don't make the developer feel like everything is saved.

File: packages/xdl/src/credentials/AndroidCredentials.js [hunk 26]
```diff
@@ -0,0 +1,255 @@
...
+      .update(data)
+      .digest('base64');
+    log(`Google Certificate Fingerprint:     ${googleHash.replace(/(.{2}(?!$))/g, '$1:')}`);
+    log(`Google Certificate Hash (SHA-1):    ${googleHash}`);
+    log(`Google Certificate Hash (SHA-256):  ${googleHash256}`);
+    log(`Facebook Key Hash:                  ${fbHash}`);
+  } catch (err) {
+    if (err.code === 'ENOENT') {
+      log.warn('Are you sure you have keytool installed?');
+      log('keytool is part of openJDK: http://openjdk.java.net/');
```
Comment (ide) [inline packages/xdl/src/credentials/AndroidCredentials.js]:
```suggestion
      log('keytool is part of OpenJDK: https://openjdk.java.net/');
```

**Pull request merged**